mtc.go

```diff
@@ -2039,3 +2044,68 @@ func (tai *TrustAnchorIdentifier) unmarshal(s *cryptobyte.String) error {
 	tai.Issuer = oid
 	return nil
 }
+
+// VerifyOpts includes parameters for verifying a BikeshedCertificate.
+type VerifyOptions struct {
+	// SignedValidityWindow is a signed validity window that covers the
+	// certificate. It is the caller's responsibility to verify the
+	// validity window's signature using the CA parameters.
+	SignedValidityWindow *SignedValidityWindow
+
+	// CA includes the parameters of the CA that issued the batch
+	// containing the certificate.
+	CA *CAParams
+
+	// CurrentTime is used to to check if the certificate has expired.
+	CurrentTime time.Time
+}
+
+// Verify is used to verify that a BikeshedCertificate is covered by a signed
+// validity window. It is the caller's responsibility to verify the validity
+// window's signature. An error indicates that the certificate does not belong
+// to a batch in the validity window or that the certificate is otherwise
+// invalid.
+func Verify(cert *BikeshedCertificate, opts VerifyOptions) error {
+	p, ok := cert.Proof.(*MerkleTreeProof)
+	if !ok {
+		return fmt.Errorf("Expected MerkleTreeProof; got %t", cert.Proof)
+	}
+
+	// Check if the certificate has expired.
+	if !opts.CurrentTime.Before(cert.Proof.NotAfter()) {
+		return fmt.Errorf("Certificate has expired (%v)", p.NotAfter().UTC())
+	}
+
+	// Check that the batch number is in range of the validity window.
+	maxBatchNumber := int(opts.SignedValidityWindow.BatchNumber)
+	minBatchNumber := maxBatchNumber - int(opts.CA.ValidityWindowSize)
+	if minBatchNumber < 0 {
+		minBatchNumber = 0
+	}
+	certBatchNumber := int(p.TrustAnchorIdentifier().BatchNumber)
+	if certBatchNumber < minBatchNumber || certBatchNumber > maxBatchNumber {
+		return fmt.Errorf("Certificate batch number (%d) out of range [%d, %d]",
+			certBatchNumber, minBatchNumber, maxBatchNumber)
+	}
+
+	// Check that the certificate issuer matches the CA.
+	if !bytes.Equal(p.TrustAnchorIdentifier().Issuer, opts.CA.Issuer) {
+		return fmt.Errorf("Certificate issuer (%s) does not match CA (%s)",
+			p.TrustAnchorIdentifier().Issuer, opts.CA.Issuer)
+	}
+
+	// Select the tree head and verify the authentication path.
+	rootIndex := int(HashLen * (maxBatchNumber - certBatchNumber))
+	root := opts.SignedValidityWindow.TreeHeads[rootIndex : rootIndex+HashLen]
+	be := NewBatchEntry(cert.Assertion, p.NotAfter())
+	b := Batch{
+		CA:     opts.CA,
+		Number: uint32(certBatchNumber),
+	}
+	err := b.VerifyAuthenticationPath(p.Index(), p.Path(), root, &be)
+	if err != nil {
+		return err
+	}
+
```

Comment:
Anything else we need to check?

Reply:
I think that's it!

Reply:
You need to check whether `CurrentTime` is within the validity window. You checked against NotAfter, but not against "NotBefore".

Reply:
Oh that's my fault--Chris asked if there's a notion of 'not before' and I said no. I guess we'd compare against the issuance time for that batch.

Reply:
I mean, I suppose we could make it the responsibility of the caller to check that the ValidityWindow is currently valid.

Reply:
Done, but please check my work:
```go
notBefore := time.Unix(int64(opts.CA.StartTime), 0).Add(
    time.Duration(uint64(certBatchNumber)*opts.CA.BatchDuration) *
        time.Second)
```